docker.buildArg.* properties not taken into account in OpenShift plugins

### Describe the bug


The JKube Openshift Maven plugin fails to recognize specified Docker build-arg when they are injected into a Dockerfile from Maven properties. This issue manifests specifically when using Maven properties and does not occur with the XML/Gradle DSL configuration since https://github.com/eclipse/jkube/pull/2895.
This issue occurs despite the JKube [documentation](https://eclipse.dev/jkube/docs/openshift-maven-plugin/#_build_args) confirming that build arguments can be specified either via the plugin configuration or as Maven properties.

### Eclipse JKube version

SNAPSHOT

### Component

OpenShift Maven Plugin

### Apache Maven version

None

### Gradle version

None

### Steps to reproduce

1. **Set Up the Dockerfile:**
   Place the following content in `src/main/docker/Dockerfile`:
   ```dockerfile
   ARG FILE_EXTENSION
   FROM docker.io/alpine:latest as builder
   RUN touch test.${FILE_EXTENSION}

   FROM docker.io/alpine:latest
   COPY --from=builder /test.txt /test.txt

   ENTRYPOINT ["ls"]
   ```

2. **Configure the JKube Plugin:**
   Insert the JKube Openshift Maven plugin configuration into your `pom.xml`:
   ```xml
   <plugin>
       <groupId>org.eclipse.jkube</groupId>
       <artifactId>openshift-maven-plugin</artifactId>
       <version>1.17-SNAPSHOT</version>
       <configuration>
         <buildStrategy>docker</buildStrategy>
         <images>
           <image>
             <name>helloworld:%l</name>
             <build>
               <contextDir>${project.basedir}/src/main/docker/</contextDir>
               <dockerFile>Dockerfile</dockerFile>
             </build>
           </image>
         </images>
       </configuration>
   </plugin>
   ```

3. **Define the Build Argument:**
   Define the build argument for the Docker image in your `pom.xml`:
   ```xml
   <properties>
       <docker.buildArg.FILE_EXTENSION>txt</docker.buildArg.FILE_EXTENSION>
   </properties>
   ```

4. **Execute the Build Command:**
   Execute the build using the following Maven command:
   ```
   mvn oc:build
   ```

5. **Error Observation:**
   The build process fails, yielding the following output:
   ```
   [INFO] oc: --> abd4c9e7124f
   [INFO] oc: [1/2] STEP 3/3: RUN touch test.${FILE_EXTENSION}
   [INFO] oc: --> ee9fcbe057df
   [INFO] oc: [2/2] STEP 1/5: FROM docker.io/alpine
   [INFO] oc: [2/2] STEP 2/5: COPY --from=builder test.txt test.txt
   [ERROR] oc: Failed to execute the build [Unable to build the image using the OpenShift build service]
   [ERROR] oc: Failed to tail build log: %s: java.lang.IllegalStateException: java.io.IOException: channel already closed with exception
   ```



### Expected behavior

The `docker.buildArg.FILE_EXTENSION` property should be properly processed, enabling the substitution of `${FILE_EXTENSION}` in the Dockerfile and ensuring the successful build of the Docker image with the specified configurations.

### Runtime

OpenShift

### Kubernetes API Server version

1.25.3

### Environment

Linux

### Eclipse JKube Logs

_No response_

### Sample Reproducer Project

_No response_

### Additional context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,6 +29,7 @@ Usage:
 * Fix #2860: Correctly pass Docker build-arg from the build configuration to the Openshift build strategy
 * Fix #2885: Provide a way to set labels on images defined by Generators
 * Fix #2901: Ensure Docker build arguments from properties are used during images pre-pulling
+* Fix #2904: `docker.buildArg.*` properties not taken into account in OpenShift plugins
 
 ### 1.16.2 (2024-03-27)
 * Fix #2461: `k8s:watch`/`k8sWatch` should throw error in `buildpacks` build strategy

--- a/jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/BuildArgResolverUtil.java
+++ b/jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/BuildArgResolverUtil.java
@@ -44,24 +44,46 @@ public class BuildArgResolverUtil {
    * @param configuration {@link JKubeConfiguration}.
    * @return a Map containing merged Build Args from all sources.
    */
-  public static Map<String, String> mergeBuildArgs(ImageConfiguration imageConfig, JKubeConfiguration configuration) {
+  public static Map<String, String> mergeBuildArgsIncludingLocalDockerConfigProxySettings(ImageConfiguration imageConfig, JKubeConfiguration configuration) {
+    return mergeBuildArgsFrom(imageConfig.getBuild().getArgs(),
+            buildArgsFromProperties(System.getProperties()),
+            buildArgsFromProperties(configuration.getProject().getProperties()),
+            configuration.getBuildArgs(),
+            buildArgsFromDockerConfig());
+  }
+
+  /**
+   * Merges Docker Build Args from the following sources:
+   * <ul>
+   *   <li>Build Args specified directly in ImageConfiguration</li>
+   *   <li>Build Args specified via System Properties</li>
+   *   <li>Build Args specified via Project Properties</li>
+   *   <li>Build Args specified via Plugin configuration</li>
+   * </ul>
+   * @param imageConfig ImageConfiguration where to get the Build Args from.
+   * @param configuration {@link JKubeConfiguration}.
+   * @return a Map containing merged Build Args from all sources.
+   */
+  public static Map<String, String> mergeBuildArgsWithoutLocalDockerConfigProxySettings(ImageConfiguration imageConfig, JKubeConfiguration configuration) {
+    return mergeBuildArgsFrom(imageConfig.getBuild().getArgs(),
+        buildArgsFromProperties(System.getProperties()),
+        buildArgsFromProperties(configuration.getProject().getProperties()),
+        configuration.getBuildArgs());
+  }
+
+  @SafeVarargs
+  private static Map<String, String> mergeBuildArgsFrom(Map<String, String>... buildArgSources) {
     final Map<String, String> buildArgs = new HashMap<>();
-    Stream.of(
-      imageConfig.getBuild().getArgs(),
-      buildArgsFromProperties(System.getProperties()),
-      buildArgsFromProperties(configuration.getProject().getProperties()),
-      configuration.getBuildArgs(),
-      buildArgsFromDockerConfig()
-    )
-      .filter(Objects::nonNull)
-      .flatMap(map -> map.entrySet().stream())
-      .forEach(entry -> {
-        if (buildArgs.containsKey(entry.getKey())) {
-          throw new JKubeException(String.format("Multiple Build Args with the same key: %s=%s and %s=%s",
-            entry.getKey(), buildArgs.get(entry.getKey()), entry.getKey(), entry.getValue()));
-        }
-        buildArgs.put(entry.getKey(), entry.getValue());
-      });
+    Stream.of(buildArgSources)
+        .filter(Objects::nonNull)
+        .flatMap(map -> map.entrySet().stream())
+        .forEach(entry -> {
+          if (buildArgs.containsKey(entry.getKey())) {
+            throw new JKubeException(String.format("Multiple Build Args with the same key: %s=%s and %s=%s",
+                entry.getKey(), buildArgs.get(entry.getKey()), entry.getKey(), entry.getValue()));
+          }
+          buildArgs.put(entry.getKey(), entry.getValue());
+        });
     return buildArgs;
   }
 

--- a/jkube-kit/build/api/src/test/java/org/eclipse/jkube/kit/build/api/helper/BuildArgResolverUtilMergeBuildArgsTest.java
+++ b/jkube-kit/build/api/src/test/java/org/eclipse/jkube/kit/build/api/helper/BuildArgResolverUtilMergeBuildArgsTest.java
@@ -22,6 +22,7 @@ import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
@@ -74,7 +75,7 @@ class BuildArgResolverUtilMergeBuildArgsTest {
         .build();
 
     // When
-    Map<String, String> mergedBuildArgs = BuildArgResolverUtil.mergeBuildArgs(imageConfiguration, jKubeConfiguration);
+    Map<String, String> mergedBuildArgs = BuildArgResolverUtil.mergeBuildArgsIncludingLocalDockerConfigProxySettings(imageConfiguration, jKubeConfiguration);
 
     // Then
     assertThat(mergedBuildArgs)
@@ -94,7 +95,7 @@ class BuildArgResolverUtilMergeBuildArgsTest {
     givenBuildArgsFromJKubeConfiguration("FULL_IMAGE", "busybox:latest");
 
     // When
-    Map<String, String> mergedBuildArgs = BuildArgResolverUtil.mergeBuildArgs(imageConfiguration, jKubeConfiguration);
+    Map<String, String> mergedBuildArgs = BuildArgResolverUtil.mergeBuildArgsIncludingLocalDockerConfigProxySettings(imageConfiguration, jKubeConfiguration);
 
     // Then
     assertThat(mergedBuildArgs)
@@ -113,7 +114,7 @@ class BuildArgResolverUtilMergeBuildArgsTest {
 
     // When & Then
     assertThatExceptionOfType(JKubeException.class)
-        .isThrownBy(() -> BuildArgResolverUtil.mergeBuildArgs(imageConfiguration, jKubeConfiguration))
+        .isThrownBy(() -> BuildArgResolverUtil.mergeBuildArgsIncludingLocalDockerConfigProxySettings(imageConfiguration, jKubeConfiguration))
         .withMessage("Multiple Build Args with the same key: VERSION=latest and VERSION=1.0.0");
   }
 
@@ -126,7 +127,7 @@ class BuildArgResolverUtilMergeBuildArgsTest {
 
     // When & Then
     assertThatExceptionOfType(JKubeException.class)
-        .isThrownBy(() -> BuildArgResolverUtil.mergeBuildArgs(imageConfiguration, jKubeConfiguration))
+        .isThrownBy(() -> BuildArgResolverUtil.mergeBuildArgsIncludingLocalDockerConfigProxySettings(imageConfiguration, jKubeConfiguration))
         .withMessage("Multiple Build Args with the same key: VERSION=latest and VERSION=1.0.0");
   }
 
@@ -139,31 +140,55 @@ class BuildArgResolverUtilMergeBuildArgsTest {
 
     // When & Then
     assertThatExceptionOfType(JKubeException.class)
-        .isThrownBy(() -> BuildArgResolverUtil.mergeBuildArgs(imageConfiguration, jKubeConfiguration))
+        .isThrownBy(() -> BuildArgResolverUtil.mergeBuildArgsIncludingLocalDockerConfigProxySettings(imageConfiguration, jKubeConfiguration))
         .withMessage("Multiple Build Args with the same key: VERSION=latest and VERSION=1.0.0");
   }
 
-  @Test
-  @DisplayName("should add proxy build args from ~/.docker/config.json")
-  void shouldAddBuildArgsFromDockerConfig(@TempDir File temporaryFolder) throws IOException {
-    try {
-      // Given
+  @Nested
+  @DisplayName("local ~/.docker/config.json contains proxy settings")
+  class LocalDockerConfigContainsProxySettings {
+    @TempDir
+    private File temporaryFolder;
+
+    @BeforeEach
+    void setUp() throws IOException {
       Path dockerConfig = temporaryFolder.toPath();
       final Map<String, String> env = Collections.singletonMap("DOCKER_CONFIG", dockerConfig.toFile().getAbsolutePath());
       EnvUtil.overrideEnvGetter(env::get);
       Files.write(dockerConfig.resolve("config.json"), ("{\"proxies\": {\"default\": {\n" +
-              "     \"httpProxy\": \"http://proxy.example.com:3128\",\n" +
-              "     \"httpsProxy\": \"https://proxy.example.com:3129\",\n" +
-              "     \"noProxy\": \"*.test.example.com,.example.org,127.0.0.0/8\"\n" +
-              "   }}}").getBytes());
+          "     \"httpProxy\": \"http://proxy.example.com:3128\",\n" +
+          "     \"httpsProxy\": \"https://proxy.example.com:3129\",\n" +
+          "     \"noProxy\": \"*.test.example.com,.example.org,127.0.0.0/8\"\n" +
+          "   }}}").getBytes());
+
+    }
+
+    @Test
+    @DisplayName("mergeBuildArgsIncludingLocalDockerConfigProxySettings, should add proxy build args for docker build strategy")
+    void shouldAddBuildArgsFromDockerConfigInDockerBuild() {
       // When
-      final Map<String, String> mergedBuildArgs = BuildArgResolverUtil.mergeBuildArgs(imageConfiguration, jKubeConfiguration);
+      final Map<String, String> mergedBuildArgs = BuildArgResolverUtil.mergeBuildArgsIncludingLocalDockerConfigProxySettings(imageConfiguration, jKubeConfiguration);
       // Then
       assertThat(mergedBuildArgs)
           .containsEntry("docker.buildArg.http_proxy", "http://proxy.example.com:3128")
           .containsEntry("docker.buildArg.https_proxy", "https://proxy.example.com:3129")
           .containsEntry("docker.buildArg.no_proxy", "*.test.example.com,.example.org,127.0.0.0/8");
-    } finally {
+    }
+
+    @Test
+    @DisplayName("mergeBuildArgsWithoutIncludingLocalDockerConfigProxySettings, should not add proxy build args for OpenShift build strategy")
+    void shouldNotAddBuildArgsFromDockerConfig() {
+      // When
+      final Map<String, String> mergedBuildArgs = BuildArgResolverUtil.mergeBuildArgsWithoutLocalDockerConfigProxySettings(imageConfiguration, jKubeConfiguration);
+      // Then
+      assertThat(mergedBuildArgs)
+          .doesNotContainEntry("docker.buildArg.http_proxy", "http://proxy.example.com:3128")
+          .doesNotContainEntry("docker.buildArg.https_proxy", "https://proxy.example.com:3129")
+          .doesNotContainEntry("docker.buildArg.no_proxy", "*.test.example.com,.example.org,127.0.0.0/8");
+    }
+
+    @AfterEach
+    void tearDown() {
       EnvUtil.overrideEnvGetter(System::getenv);
     }
   }

--- a/jkube-kit/build/service/docker/src/main/java/org/eclipse/jkube/kit/build/service/docker/BuildService.java
+++ b/jkube-kit/build/service/docker/src/main/java/org/eclipse/jkube/kit/build/service/docker/BuildService.java
@@ -34,7 +34,7 @@ import org.eclipse.jkube.kit.config.image.ImageName;
 import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
 import org.eclipse.jkube.kit.config.image.build.CleanupMode;
 
-import static org.eclipse.jkube.kit.build.api.helper.BuildArgResolverUtil.mergeBuildArgs;
+import static org.eclipse.jkube.kit.build.api.helper.BuildArgResolverUtil.mergeBuildArgsIncludingLocalDockerConfigProxySettings;
 import static org.eclipse.jkube.kit.build.api.helper.BuildUtil.extractBaseFromConfiguration;
 
 public class BuildService {
@@ -64,7 +64,7 @@ public class BuildService {
     public void buildImage(ImageConfiguration imageConfig, ImagePullManager imagePullManager, JKubeConfiguration configuration)
             throws IOException {
 
-        Map<String, String> mergedBuildArgs = mergeBuildArgs(imageConfig, configuration);
+        Map<String, String> mergedBuildArgs = mergeBuildArgsIncludingLocalDockerConfigProxySettings(imageConfig, configuration);
 
         if (imagePullManager != null) {
             autoPullBaseImage(imageConfig, imagePullManager, configuration, mergedBuildArgs);

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/openshift/OpenShiftBuildServiceUtils.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/openshift/OpenShiftBuildServiceUtils.java
@@ -47,6 +47,7 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.stream.Collectors;
 
+import static org.eclipse.jkube.kit.build.api.helper.BuildArgResolverUtil.mergeBuildArgsWithoutLocalDockerConfigProxySettings;
 import static org.eclipse.jkube.kit.build.api.helper.BuildUtil.extractBaseFromDockerfile;
 import static org.eclipse.jkube.kit.config.service.openshift.ImageStreamService.resolveImageStreamName;
 import static org.eclipse.jkube.kit.config.service.openshift.OpenshiftBuildService.DEFAULT_BUILD_OUTPUT_KIND;
@@ -133,7 +134,9 @@ public class OpenShiftBuildServiceUtils {
               .withNamespace(StringUtils.isEmpty(fromNamespace) ? null : fromNamespace)
             .endFrom()
             .withEnv(checkForEnv(imageConfig))
-            .withBuildArgs(Optional.ofNullable(buildConfig.getArgs()).orElse(Collections.emptyMap()).entrySet().stream()
+            .withBuildArgs(mergeBuildArgsWithoutLocalDockerConfigProxySettings(imageConfig, jKubeServiceHub.getConfiguration())
+                .entrySet()
+                .stream()
                 .map(bcArg -> new EnvVarBuilder()
                     .withName(bcArg.getKey())
                     .withValue(bcArg.getValue()).build())

--- a/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenShiftBuildServiceUtilsTest.java
+++ b/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenShiftBuildServiceUtilsTest.java
@@ -18,12 +18,15 @@ import java.io.IOException;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Properties;
 
 import io.fabric8.openshift.api.model.ImageStreamTag;
 import io.fabric8.openshift.api.model.ImageStreamTagBuilder;
+import org.assertj.core.api.InstanceOfAssertFactories;
 import org.eclipse.jkube.kit.build.api.assembly.ArchiverCustomizer;
 import org.eclipse.jkube.kit.build.api.assembly.JKubeBuildTarArchiver;
 import org.eclipse.jkube.kit.common.JKubeConfiguration;
+import org.eclipse.jkube.kit.common.JavaProject;
 import org.eclipse.jkube.kit.config.image.ImageConfiguration;
 import org.eclipse.jkube.kit.config.image.ImageName;
 import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
@@ -40,6 +43,8 @@ import io.fabric8.openshift.api.model.BuildOutput;
 import io.fabric8.openshift.api.model.BuildStrategy;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 import org.mockito.ArgumentCaptor;
@@ -75,6 +80,10 @@ class OpenShiftBuildServiceUtilsTest {
     jKubeServiceHub = mock(JKubeServiceHub.class, RETURNS_DEEP_STUBS);
     when(jKubeServiceHub.getBuildServiceConfig().getBuildDirectory())
         .thenReturn(temporaryFolder.getAbsolutePath());
+    when(jKubeServiceHub.getConfiguration()).thenReturn(JKubeConfiguration.builder()
+            .project(JavaProject.builder().build())
+            .buildArgs(Collections.emptyMap())
+        .build());
     imageConfiguration = ImageConfiguration.builder()
         .name("myapp")
         .build(BuildConfiguration.builder()
@@ -183,64 +192,95 @@ class OpenShiftBuildServiceUtilsTest {
         .hasFieldOrPropertyWithValue("pullSecret.name", "my-secret-for-pull");
   }
 
-  @Test
-  void createBuildStrategy_withDockerBuildStrategyAndNoPullSecret_shouldReturnValidBuildStrategy() {
-    // Given
-    when(jKubeServiceHub.getBuildServiceConfig().getJKubeBuildStrategy())
-        .thenReturn(JKubeBuildStrategy.docker);
-    // When
-    final BuildStrategy result = createBuildStrategy(jKubeServiceHub, imageConfiguration, null);
-    // Then
-    assertThat(result)
-        .hasFieldOrPropertyWithValue("type", "Docker")
-        .extracting(BuildStrategy::getDockerStrategy)
-        .hasFieldOrPropertyWithValue("from.kind", "DockerImage")
-        .hasFieldOrPropertyWithValue("from.name", "ubi8/s2i-base");
-  }
+  @Nested
+  @DisplayName("docker BuildStrategy")
+  class DockerBuildStrategy {
+    @BeforeEach
+    void setUp() {
+      when(jKubeServiceHub.getBuildServiceConfig().getJKubeBuildStrategy())
+          .thenReturn(JKubeBuildStrategy.docker);
+    }
 
-  @Test
-  void createBuildStrategy_withDockerBuildArgs_shouldReturnValidBuildStrategyWithBuildArgs() {
-    // Given
-    when(jKubeServiceHub.getBuildServiceConfig().getJKubeBuildStrategy())
-        .thenReturn(JKubeBuildStrategy.docker);
-    ImageConfiguration imageConfig = ImageConfiguration.builder()
-        .name("myapp")
-        .build(BuildConfiguration.builder()
-            .from("ubi8/s2i-base")
-            .args(Collections.singletonMap("BUILD_ARGS_KEY", "build-args-value"))
-            .build())
-        .build();
-    // When
-    final BuildStrategy result = createBuildStrategy(jKubeServiceHub, imageConfig, "my-secret-for-pull");
-    // Then
-    assertThat(result)
-        .hasFieldOrPropertyWithValue("type", "Docker")
-        .extracting(BuildStrategy::getDockerStrategy)
-        .hasFieldOrPropertyWithValue("from.kind", "DockerImage")
-        .hasFieldOrPropertyWithValue("from.name", "ubi8/s2i-base")
-        .extracting("buildArgs").asList().hasSize(1).first().satisfies(
-            envVar -> {
-              assertThat(envVar)
-                  .hasFieldOrPropertyWithValue("name", "BUILD_ARGS_KEY")
-                  .hasFieldOrPropertyWithValue("value", "build-args-value");
-            });
-  }
+    @Test
+    @DisplayName("no pull secret specified")
+    void withDockerBuildStrategyAndNoPullSecret_shouldReturnValidBuildStrategy() {
+      // When
+      final BuildStrategy result = createBuildStrategy(jKubeServiceHub, imageConfiguration, null);
+      // Then
+      assertThat(result)
+          .hasFieldOrPropertyWithValue("type", "Docker")
+          .extracting(BuildStrategy::getDockerStrategy)
+          .hasFieldOrPropertyWithValue("from.kind", "DockerImage")
+          .hasFieldOrPropertyWithValue("from.name", "ubi8/s2i-base");
+    }
 
-  @Test
-  void createBuildStrategy_withDockerBuildStrategyAndPullSecret_shouldReturnValidBuildStrategy() {
-    // Given
-    when(jKubeServiceHub.getBuildServiceConfig().getJKubeBuildStrategy())
-        .thenReturn(JKubeBuildStrategy.docker);
-    // When
-    final BuildStrategy result = createBuildStrategy(jKubeServiceHub, imageConfiguration, "my-secret-for-pull");
-    // Then
-    assertThat(result)
-        .hasFieldOrPropertyWithValue("type", "Docker")
-        .extracting(BuildStrategy::getDockerStrategy)
-        .hasFieldOrPropertyWithValue("from.kind", "DockerImage")
-        .hasFieldOrPropertyWithValue("from.name", "ubi8/s2i-base")
-        .hasFieldOrPropertyWithValue("noCache", false)
-        .hasFieldOrPropertyWithValue("pullSecret.name", "my-secret-for-pull");
+    @Test
+    @DisplayName("pull secret specified, should reflect in Build Strategy")
+    void withPullSecret_shouldReturnValidBuildStrategy() {
+      // When
+      final BuildStrategy result = createBuildStrategy(jKubeServiceHub, imageConfiguration, "my-secret-for-pull");
+      // Then
+      assertThat(result)
+          .hasFieldOrPropertyWithValue("type", "Docker")
+          .extracting(BuildStrategy::getDockerStrategy)
+          .hasFieldOrPropertyWithValue("from.kind", "DockerImage")
+          .hasFieldOrPropertyWithValue("from.name", "ubi8/s2i-base")
+          .hasFieldOrPropertyWithValue("noCache", false)
+          .hasFieldOrPropertyWithValue("pullSecret.name", "my-secret-for-pull");
+    }
+
+    @Nested
+    @DisplayName("with build args")
+    class WithDockerBuildArgs {
+      @Test
+      @DisplayName("when docker build args in ImageConfig, then add them to BuildStrategy")
+      void withDockerBuildArgsInImageConfiguration_shouldReturnValidBuildStrategyWithBuildArgs() {
+        // Given
+        imageConfiguration = imageConfiguration.toBuilder()
+            .build(imageConfiguration.getBuild().toBuilder()
+                .args(Collections.singletonMap("BUILD_ARGS_KEY", "build-args-value"))
+                .build())
+            .build();
+        // When
+        final BuildStrategy result = createBuildStrategy(jKubeServiceHub, imageConfiguration, "my-secret-for-pull");
+        // Then
+        assertBuildArgPresentInBuildStrategy(result);
+      }
+
+      @Test
+      @DisplayName("when docker build args in project properties, then add them to BuildStrategy")
+      void withDockerBuildArgsInProjectProperties_shouldReturnValidBuildStrategyWithBuildArgs() {
+        // Given
+        Properties properties = new Properties();
+        properties.put("docker.buildArg.BUILD_ARGS_KEY", "build-args-value");
+        when(jKubeServiceHub.getConfiguration()).thenReturn(JKubeConfiguration.builder()
+            .project(JavaProject.builder()
+                .properties(properties)
+                .build())
+            .build());
+        // When
+        final BuildStrategy result = createBuildStrategy(jKubeServiceHub, imageConfiguration, "my-secret-for-pull");
+        // Then
+        assertBuildArgPresentInBuildStrategy(result);
+      }
+
+      void assertBuildArgPresentInBuildStrategy(BuildStrategy result) {
+        assertThat(result)
+            .hasFieldOrPropertyWithValue("type", "Docker")
+            .extracting(BuildStrategy::getDockerStrategy)
+            .hasFieldOrPropertyWithValue("from.kind", "DockerImage")
+            .hasFieldOrPropertyWithValue("from.name", "ubi8/s2i-base")
+            .extracting("buildArgs")
+            .asInstanceOf(InstanceOfAssertFactories.LIST)
+            .singleElement()
+            .satisfies(
+                envVar -> {
+                  assertThat(envVar)
+                      .hasFieldOrPropertyWithValue("name", "BUILD_ARGS_KEY")
+                      .hasFieldOrPropertyWithValue("value", "build-args-value");
+                });
+      }
+    }
   }
 
   @Test

